### PR TITLE
[Merged by Bors] - feat(CategoryTheory/Enriched): Add eHomCongr and related lemmas

### PR DESCRIPTION
**Pull request by gio256**
Recall that when `C` is both a category and a `V`-enriched category, we say it is an `EnrichedOrdinaryCategory` if it comes equipped with an equivalence between morphisms `X ⟶ Y` in `C`  and morphisms `𝟙_ V ⟶ (X ⟶[V] Y)` in `V` that preserves identities and composition in `C`.

In such a `V`-enriched ordinary category `C`, isomorphisms in `C` induce isomorphisms between hom-objects in `V`. We define this isomorphism, `CategoryTheory.Iso.eHomCongr`, and prove that it behaves nicely with respect to composition in `C`.

The lemmas here parallel those for unenriched categories in [`Mathlib/CategoryTheory/HomCongr.lean`](https://github.com/leanprover-community/mathlib4/blob/a1c80a59b69da977f8519230ef96b3b95e31bc1d/Mathlib/CategoryTheory/HomCongr.lean) and those for sorts in [`Mathlib/Logic/Equiv/Defs.lean`](https://github.com/leanprover-community/mathlib4/blob/a1c80a59b69da977f8519230ef96b3b95e31bc1d/Mathlib/Logic/Equiv/Defs.lean#L428) (cf. `Equiv.arrowCongr`). Note, however, that they construct equivalences between `Type`s and `Sort`s, respectively, while in this PR we construct isomorphisms between objects in `V`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Enriched/HomCongr.lean [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+  hom_inv_id := by
+    rw [← eHom_whisker_exchange, assoc]
+    rw [← eHomWhiskerRight_comp_assoc, hom_inv_id, eHomWhiskerRight_id, id_comp]
+    rw [← eHomWhiskerLeft_comp, hom_inv_id, eHomWhiskerLeft_id]
+  inv_hom_id := by
+    rw [← eHom_whisker_exchange, assoc]
+    rw [← eHomWhiskerRight_comp_assoc, inv_hom_id, eHomWhiskerRight_id, id_comp]
+    rw [← eHomWhiskerLeft_comp, inv_hom_id, eHomWhiskerLeft_id]
+
+lemma eHomCongr_refl {X Y : C} :
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Enriched/HomCongr.lean:47]:
```suggestion
lemma eHomCongr_refl (X Y : C) :
```

File: Mathlib/CategoryTheory/Enriched/HomCongr.lean [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  inv := eHomWhiskerRight V α.hom Y₁ ≫ eHomWhiskerLeft V X β.inv
+  hom_inv_id := by
+    rw [← eHom_whisker_exchange, assoc]
+    rw [← eHomWhiskerRight_comp_assoc, hom_inv_id, eHomWhiskerRight_id, id_comp]
+    rw [← eHomWhiskerLeft_comp, hom_inv_id, eHomWhiskerLeft_id]
+  inv_hom_id := by
+    rw [← eHom_whisker_exchange, assoc]
+    rw [← eHomWhiskerRight_comp_assoc, inv_hom_id, eHomWhiskerRight_id, id_comp]
+    rw [← eHomWhiskerLeft_comp, inv_hom_id, eHomWhiskerLeft_id]
+
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/Enriched/HomCongr.lean:46]:
```suggestion
  hom_inv_id := by
    rw [← eHom_whisker_exchange]
    slice_lhs 2 3 => rw [← eHomWhiskerRight_comp]
    simp [← eHomWhiskerLeft_comp]
  inv_hom_id := by
    rw [← eHom_whisker_exchange]
    slice_lhs 2 3 => rw [← eHomWhiskerRight_comp]
    simp [← eHomWhiskerLeft_comp]

```
There is nothing wrong with your proofs, this is just a suggestion that shows how you can let `simp` handle the trivial stuff like `id_comp`, `assoc`, `hom_inv_id`, etc.

File: Mathlib/CategoryTheory/Enriched/HomCongr.lean [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        _ ◁ (eHomEquiv V g ≫ (eHomCongr V β γ).hom) ≫ eComp V X₁ Y₁ Z₁ := by
+  dsimp only [eHomCongr]
+  simp only [assoc, MonoidalCategory.whiskerRight_id,
+    MonoidalCategory.whiskerLeft_comp]
+  rw [rightUnitor_inv_naturality_assoc, rightUnitor_inv_naturality_assoc,
+    rightUnitor_inv_naturality_assoc, hom_inv_id_assoc]
+  rw [← whisker_exchange_assoc, ← whisker_exchange_assoc, ← eComp_eHomWhiskerLeft]
+  rw [eHom_whisker_cancel_assoc]
+  rw [← eComp_eHomWhiskerRight_assoc, ← tensorHom_def_assoc,
+    ← eHomEquiv_comp_assoc]
```
[the remark this answers is not in the document]
Comment (gio256) [reply Mathlib/CategoryTheory/Enriched/HomCongr.lean:79]: I was curious about this, actually. I like being able to split up logical steps of the proof within a long string of rewrites, but I also see the aesthetic advantage of combining them. I wasn't able to find anything in the style guide about it, but I do think your suggestion is more in line with what I've seen in mathlib. Thanks.

[the remark this answers is not in the document]
Comment (dagurtomas) [reply Mathlib/CategoryTheory/Enriched/HomCongr.lean:79]: Note that you can follow the logical steps of a proof by moving your cursor within a big rewrite block (you see how the infoview changes after each lemma)

File: Mathlib/CategoryTheory/Enriched/Ordinary.lean [hunk 4]
```diff
@@ -104,6 +113,35 @@ lemma eHomWhiskerLeft_comp (X : C) {Y Y' Y'' : C} (g : Y ⟶ Y') (g' : Y' ⟶ Y'
     associator_inv_naturality_right_assoc, Iso.hom_inv_id_assoc,
     whisker_exchange_assoc, MonoidalCategory.whiskerRight_id_assoc, Iso.inv_hom_id_assoc]
 
+/-- Whiskering commutes with the enriched composition. -/
+@[reassoc]
+lemma eComp_eHomWhiskerLeft (X Y : C) {Z Z' : C} (g : Z ⟶ Z') :
+    eComp V X Y Z ≫ eHomWhiskerLeft V X g =
+      _ ◁ eHomWhiskerLeft V Y g ≫ eComp V X Y Z' := by
+  unfold eHomWhiskerLeft
+  rw [rightUnitor_inv_naturality_assoc, ← whisker_exchange_assoc]
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/Enriched/Ordinary.lean:122]:
```suggestion
  dsimp [eHomWhiskerLeft]
  rw [rightUnitor_inv_naturality_assoc, ← whisker_exchange_assoc]
```

File: Mathlib/CategoryTheory/Enriched/Ordinary.lean [hunk 5]
```diff
@@ -104,6 +113,35 @@ lemma eHomWhiskerLeft_comp (X : C) {Y Y' Y'' : C} (g : Y ⟶ Y') (g' : Y' ⟶ Y'
...
+/-- Given an isomorphism `α : Y ≅ Y₁` in C, the enriched composition map
+`eComp V X Y Z : (X ⟶[V] Y) ⊗ (Y ⟶[V] Z) ⟶ (X ⟶[V] Z)` factors through the `V`
+object `(X ⟶[V] Y₁) ⊗ (Y₁ ⟶[V] Z)` via the map defined by whiskering in the
+middle with `α.hom` and `α.inv`. -/
+@[reassoc]
+lemma eHom_whisker_cancel {X Y Y₁ Z : C} (α : Y  ≅ Y₁) :
+    eHomWhiskerLeft V X α.hom ▷ _ ≫ _ ◁ eHomWhiskerRight V α.inv Z ≫
+      eComp V X Y₁ Z = eComp V X Y Z := by
+  unfold eHomWhiskerLeft eHomWhiskerRight
+  simp only [MonoidalCategory.whiskerLeft_comp_assoc, whisker_assoc_symm,
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/Enriched/Ordinary.lean:134]:
```suggestion
  dsimp [eHomWhiskerLeft, eHomWhiskerRight]
  simp only [MonoidalCategory.whiskerLeft_comp_assoc, whisker_assoc_symm,
```

File: Mathlib/CategoryTheory/Enriched/Ordinary.lean [hunk 6]
```diff
@@ -81,6 +81,15 @@ lemma eHomWhiskerRight_comp {X X' X'' : C} (f : X ⟶ X') (f' : X' ⟶ X'') (Y :
     ← associator_inv_naturality_left_assoc, Iso.inv_hom_id_assoc,
     ← whisker_exchange_assoc, id_whiskerLeft_assoc, Iso.inv_hom_id_assoc]
 
+/-- Whiskering commutes with the enriched composition. -/
+@[reassoc]
+lemma eComp_eHomWhiskerRight {X X' : C} (f : X ⟶ X') (Y Z : C) :
+    eComp V X' Y Z ≫ eHomWhiskerRight V f Z =
+      eHomWhiskerRight V f Y ▷ _ ≫ eComp V X Y Z := by
+  unfold eHomWhiskerRight
+  rw [leftUnitor_inv_naturality_assoc, whisker_exchange_assoc]
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/Enriched/Ordinary.lean:90]:
```suggestion
  dsimp [eHomWhiskerRight]
  rw [leftUnitor_inv_naturality_assoc, whisker_exchange_assoc]
```

**Review by dagurtomas — APPROVED** [review]
Thanks!

maintainer delegate

File: Mathlib/CategoryTheory/Enriched/Ordinary.lean (hunk 4 above)
Comment (gio256) [reply Mathlib/CategoryTheory/Enriched/Ordinary.lean:122]: Is there a reason you prefer `dsimp` to `unfold`? 

In Coq I believe `unfold` would be the "weakest" version of "replace this thing with it's definition", but it seems that is not the case in lean. As I understand it, `unfold` in lean is just `simp` with some special settings, but I haven't been able to figure out what those settings are.

Comment (dagurtomas) [reply Mathlib/CategoryTheory/Enriched/Ordinary.lean]: I don't think there is much difference between `unfold foo` and `dsimp [foo]`, the suggestion was just to conform with the style of the surrounding proofs. I also think the `dsimp` version is more common in mathlib, but that's just a feeling.

Comment (gio256) [reply Mathlib/CategoryTheory/Enriched/Ordinary.lean]: That's helpful, thank you.

**Comment by joelriou** [comment]
Thanks!

bors merge

**Pull request closed without merging**